### DISCUSSION TITLE
Fine Grained Personal Access Tokens - Create programmatically / via API?

### DISCUSSION THREAD

**Discussion by IliqNikushev** [community/community#120437] [Programming Help]
### Select Topic Area

Question

### Body

Hello,

I have been playing around with Personal Access Tokens and so far the Fine Grained PAT's are a pat on the back (pun intended)

They give you great control and flexibility. I am wondering if there is a way to programmatical create them?

My use case is that we have a lot of developers, teams and departments and i do not want to have Only one bot for it all
- It would be great if i can dynamically create PAT's so that they have access to specific repositories (for having a Per-Team bot)
- it would be nice to have a nice way to manage them, because creating calendar reminders "hey token X is expiring" is a bit tedious

Is there a way to programmatically create/delete Fine-Grained Personal Access Tokens? 
Thanks in advance!

**Comment by Rashwo** [comment]
> ### Select Topic Area
> Question
> 
> ### Body
> Hello,
> 
> I have been playing around with Personal Access Tokens and so far the Fine Grained PAT's are a pat on the back (pun intended)
> 
> They give you great control and flexibility. I am wondering if there is a way to programmatical create them?
> 
> My use case is that we have a lot of developers, teams and departments and i do not want to have Only one bot for it all
> 
> * It would be great if i can dynamically create PAT's so that they have access to specific repositories (for having a Per-Team bot)
> * it would be nice to have a nice way to manage them, because creating calendar reminders "hey token X is expiring" is a bit tedious
> 
> Is there a way to programmatically create/delete Fine-Grained Personal Access Tokens? Thanks in advance!

**Comment by peekcodedev** [comment]
Indeed, Personal Access Tokens (PATs) with fine-grained PATs provide greater flexibility and control compared to regular PATs. You want to create PATs programmatically, especially to support multiple developers, teams, and departments.

You can use the GitHub API to programmatically create and delete Fine-Grained PATs. Here are the general steps to do this:

### Creating Fine-Grained PATs Programmatically:

You can use `curl` or other HTTP client libraries to call the GitHub API and create PATs. Here's an example using `curl`:


curl -X POST\
      -H "Authorization: token YOUR_PERSONAL_ACCESS_TOKEN" \
      -H "Accept: application/vnd.github.v3+json" \
      https://api.github.com/repos/OWNER/REPO/actions/runners/registration-token


In the example above, `YOUR_PERSONAL_ACCESS_TOKEN` is your personal access token with sufficient privileges to create PATs. `OWNER` is the owner of the repository, and `REPO` is the name of the repository.

### Removing Fine-Grained PATs Programmatically:

You can also use the GitHub API to remove PATs. Here's an example using `curl`:


curl -X DELETE \
      -H "Authorization: token YOUR_PERSONAL_ACCESS_TOKEN" \
      https://api.github.com/repos/OWNER/REPO/actions/runners/registration-token


### PATs Management:

For your PATs management, you can create scripts or applications that perform logic to:

- Create PATs for each team or department.
- Store PATs and related information (such as expiration dates) in a database or secure repository.
- Send automatic notifications or emails when PATs are approaching their expiry date.

For more sophisticated management, you can also consider using an authentication management tool or CI/CD solution that supports automated token management.

Using this approach, you can create a dynamic and efficient solution for the management of Fine-Grained PATs in your GitHub environment.
